designs/sam_publish_app_cmd.rst

```diff
@@ -0,0 +1,229 @@
+.. contents:: **Table of Contents**
+   :depth: 2
+   :local:
+
+``sam publish app`` command
+====================================
+
+This is the design for a command to publish an application to `AWS Serverless Application Repository (SAR)`_ with a SAM
+template. It can be used to create a new application and its first version, update exisitng application's metadata, create
+a new version of the application, and manage application permissions.
+
+.. _AWS Serverless Application Repository (SAR): https://aws.amazon.com/serverless/serverlessrepo/
+
+
+What is the problem?
+--------------------
+To publish an app to AWS Serverless Application Repository, customers need to go through the following steps: first upload
+the application code and SAM template to an Amazon S3 bucket, correctly set S3 bucket policy that grants the service read
+permissions for artifacts uploaded to S3, then open the AWS Serverless Application Repository console and provide information
+in a bunch of input boxes. If they use the AWS CLI, they need to pass all the information as parameters, and it's easy to make
+a mistake while typing in the command line.
+
+
+What will be changed?
+---------------------
+In this proposal, we will be providing a new command, ``sam publish app``, which takes a SAM template as input and publishes
+an application to AWS Serverless Application Repository using applicaiton metadata specified in the template. Customers just
+need to provide application metadata information in the template, then ``sam publish app`` will handle uploading local files
+to S3 and creating the app. We will also provde sharing options to set application permission policies. This command will
+greatly simplify the exsiting publishing experience.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all the following use cases:
+
+   * Create new application and its first version in SAR using ``sam publish app``
+   * Update existing SAR application (metadata only) using ``sam publish app --meta-data``
+   * Create new version of existing SAR application using ``sam publish app``
+   * Share the app publicly using the ``--make-public`` option
+   * Make the app private using the ``--make-private`` option
+   * Share the app privately with other AWS accounts using the ``--account-ids`` option
+
+
+#. ``sam package`` command can upload local readme/license files to S3.
+
+#. sample app template generated by ``sam init`` will include the "AWS::ServerlessRepo::Application" section.
+
+
+Out-of-Scope
+------------
+#. Create new application or version if the ``--meta-data`` option is used.
```

Comment:
In SAR, application version and metadata are separate entities, so updating the metadata doesn't create a new version. Initially we thought about just using `sam publish app` to update metadata and publish app, but there are some fields that can't be modified after the app/version is created which makes it complicated. 

For example, license can't be changed after an app is created, if customers provide a different license, should we ignore it or fail the command? Ignoring the new license value may make customers feel that it has been updated because the publish command succeeds. Failing the command would require them to provide different metadata spec. 

Considering the above, we decided to separate updating metadata from publishing app.

Reply:
PR to support updating metadata with publishing and returning details if you'd like to take a look: https://github.com/awslabs/aws-serverlessrepo-python/pull/14.